dbt_common/record.py

```diff
@@ -51,31 +53,133 @@ def from_dict(cls, dct: Mapping) -> "Record":
 
 
 class Diff:
-    """Marker class for diffs?"""
+    def __init__(self, current_recording_path: str, previous_recording_path: str) -> None:
+        self.current_recording_path = current_recording_path
+        self.previous_recording_path = previous_recording_path
+
+    def diff_query_records(self, current: List, previous: List) -> Dict[str, Any]:
+        # some of the table results are returned as a stringified list of dicts that don't
+        # diff because order isn't consistent. convert it into a list of dicts so it can
+        # be diffed ignoring order
+
+        for i in range(len(current)):
+            if current[i].get("result").get("table") is not None:
+                current[i]["result"]["table"] = json.loads(current[i]["result"]["table"])
+        for i in range(len(previous)):
+            if previous[i].get("result").get("table") is not None:
+                previous[i]["result"]["table"] = json.loads(previous[i]["result"]["table"])
+
+        return DeepDiff(previous, current, ignore_order=True, verbose_level=2)
+
+    def diff_env_records(self, current: List, previous: List) -> Dict[str, Any]:
+        # The mode and filepath may change.  Ignore them.
+
+        exclude_paths = [
+            "root[0]['result']['env']['DBT_RECORDER_FILE_PATH']",
+            "root[0]['result']['env']['DBT_RECORDER_MODE']",
+        ]
+
+        return DeepDiff(
+            previous, current, ignore_order=True, verbose_level=2, exclude_paths=exclude_paths
+        )
 
-    pass
+    def diff_ignore_non_root(self, current: List, previous: List) -> Dict[str, Any]:
+        # TODO: do we want to exclude non-root from the diff to avoid diffing the
+        # contents of macros that were modified externally?  If so we'll need some
+        # metadata like the project name to know what to exclude.
+        # we need the non-root values for replay so we can't always exclude them
```

Comment:
@peterallenwebb do you have any thoughts on this?  It's not blocking so I can remove it for now.  But I'm wondering if there's other project level metadata we may need.

Reply:
removed this block!